java/vector/src/main/java/org/apache/arrow/vector/dictionary/DictionaryHashTable.java

```diff
@@ -0,0 +1,271 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.dictionary;
+
+import java.util.Objects;
+
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * HashTable used for Dictionary encoding. It holds two vectors (the vector to encode and dictionary vector)
+ * It stores the index in dictionary vector and for a given index in encode vector,
+ * it could return dictionary index.
+ */
+public class DictionaryHashTable {
+
+  /**
+   * Represents a null value in map.
+   */
+  static final int NULL_VALUE = -1;
+
+  /**
+   * The default initial capacity - MUST be a power of two.
+   */
+  static final int DEFAULT_INITIAL_CAPACITY = 1 << 4;
+
+  /**
+   * The maximum capacity, used if a higher value is implicitly specified
+   * by either of the constructors with arguments.
+   */
+  static final int MAXIMUM_CAPACITY = 1 << 30;
+
+  /**
+   * The load factor used when none specified in constructor.
+   */
+  static final float DEFAULT_LOAD_FACTOR = 0.75f;
+
+  static final DictionaryHashTable.Entry[] EMPTY_TABLE = {};
+
+  /**
+   * The table, initialized on first use, and resized as
+   * necessary. When allocated, length is always a power of two.
+   */
+  transient DictionaryHashTable.Entry[] table = (DictionaryHashTable.Entry[]) EMPTY_TABLE;
+
+  /**
+   * The number of key-value mappings contained in this map.
+   */
+  transient int size;
+
+  /**
+   * The next size value at which to resize (capacity * load factor).
+   */
+  int threshold;
+
+  /**
+   * The load factor for the hash table.
+   */
+  final float loadFactor;
+
+  private final ValueVector dictionary;
+
+  private final ValueVector toEncode;
```

Comment:
Is there a reason you chose to make this a member?  On a first pass it would make more sense to me take it as a parameter on getInde?

Reply:
Sure, toEncode should as a param in which dictionary could be used for multiple vectors, fixed.